README.md
JFreeChart
==========

Version 1.5.0 (and 1.0.20), not yet released.


Overview
--------
JFreeChart is a comprehensive free chart library for the Java(tm) platform that 
can be used on the client-side (JavaFX and Swing) or the server side (with
export to multiple formats including SVG, PNG and PDF).

The home page for the project is:

http://www.jfree.org/jfreechart

JFreeChart requires JDK 1.8.0_40 or later and is licensed under the terms of the
GNU Lesser General Public License (LGPL) version 2.1 or later.


Building JFreeChart
-------------------
You can build JFreeChart using either Ant or Maven.  The build requires JDK 1.6.0 or later. 
If you require JavaFX support, then JDK 1.8.0_40 or later is required.

#### Ant
Using Ant you can build JFreeChart using the following command (issued from the root directory of the project):

    ant -f ant/build-fx.xml

This will include JavaFX support.  To build without JavaFX:

    ant -f ant/build.xml

The Ant build script will perform all build and packaging tasks.

#### Maven
Using Maven you can build JFreeChart using the following command (issued from the root directory of the project):

    mvn clean install

This will include the JavaFX support.  To build without JavaFX, you should copy the content of the file `maven/pom-1.6.xml` into the `pom.xml` file in the root directory and then type the same command as above.

History
-------

##### Version 1.0.19 (31-Jul-2014)
- fixed clipping issues for combined plots in JavaFX;
- fixed a memory leak in the new JavaFX `ChartCanvas` class;
- `CombinedDomainXYPlot` and `CombinedRangeXYPlot` now take into account the pannable flags in the subplots;
- `FastScatterPlot` panning direction is corrected;
- added rendering hints to sharpen gridlines and borders in most output formats;        
- JFreeSVG updated to version 2.0;
- included a preview of JSFreeChart, a 2D chart library written in JavaScript that is conceptually similar to JFreeChart but runs directly in a web browser.

##### Version 1.0.18 (3-Jul-2014)
- added JavaFX support via `FXGraphics2D`;
- improved `LogAxis` labelling;
- improved numeric tick labelling;
- center text support in `RingPlot`;
- `stepPoint` attribute in the `XYStepAreaRenderer`;
- other minor improvements and bug fixes.

##### Version 1.0.17 (22-Nov-2013)
- Enhanced `XYSplineRenderer` with new area fill (contributed by Klaus Rheinwald);
- added a notify flag to all datasets that extend `AbstractDataset`;
- extended `TimeSeriesCollection` to validate `TimeSeries` keys for uniqueness;
- added a new `DirectionalGradientPaintTransformer` (by Peter Kolb);
- updated `OHLCSeries`;
- added `HMSNumberFormat`;
- updated JCommon to version 1.0.21 (includes rotated text improvements) and fixed some minor bugs.

###### Bug Fixes
- 977  : Multithreading issue with `DateAxis`;
- 1084 : `BorderArrangement.add()` possible `ClassCastException`;
- 1099 : `XYSeriesCollection.removeSeries(int)` does not deregister listener;
- 1109 : `WaterfallBarRenderer` uses wrong color for diff 0.


##### Version 1.0.16 (13-Sep-2013)

*** THIS RELEASE REQUIRES JDK/JRE 1.6.0 OR LATER. ***

- Provided subscript/superscript support for axis labels (via AttributedString);
- new axis label positioning options;
- simplified ChartFactory methods;
- added new methods to DatasetUtilities to interpolate y-values in XYDatasets;
- added URLs to labels on CategoryAxis;
- seamless integration with JFreeSVG (http://www.jfree.org/jfreesvg/) and OrsonPDF 
(http://www.object-refinery.com/pdf/);
- improved the consistency of the SWTGraphics2D implementatio;  

All the JUnit tests have been upgraded to JUnit 4.

###### Bug Fixes

- 1107 : Fixed TimeZone issue in PeriodAxis;

Also fixed a line drawing issue with the StackedXYAreaRenderer, and a memory 
leak in the SWT ChartComposite class.


##### Version 1.0.15 (4-Jul-2013)
- Added support for non-visible series in XYBarRenderer;
- minor gridlines in PolarPlot;
- legend item ordering;
- chart editor enhancements;
- updates to StandardDialScale;
- localisation files for Japanese;
- refactored parameter checks.  

This release also fixes a minor security flaw in the DisplayChart class, detected and reported by OSI Security:

http://www.osisecurity.com.au/advisories/jfreechart-path-disclosure

###### Patches

- 3500621 : LegendTitle order attribute (by Simon Kaczor);
- 3463807 : ChartComposite does not dispose popup (by Sebastiao Correia);
- 3204823 : PaintAlpha for 3D effects (by Dave Law);

###### Bug Fixes

- 3561093 : Rendering anomaly for XYPlots;
- 3555275 : ValueAxis.reserveSpace() problem for axes with fixed dimension;
- 3521736 : DeviationRenderer optimisation (by Milan Ramaiya);
- 3514487 : SWTGraphics2D get/setStroke() problem;
- 3508799 : DefaultPolarItemRenderer does not populate seriesKey in LegendItem;
- 3482106 : Missing text in SWTGraphics2D (by Kevin Xu);
- 3484408 : Maven fixes (Martin Hoeller);
- 3484403 : DateAxis endless loop (by Martin Hoeller);
- 3446965 : TimeSeries calculates range incorrectly in addOrUpdate();
- 3445507 : TimeSeriesCollection.findRangeBounds() regression;
- 3425881 : XYDifferenceRenderer fix (by Patrick Schlott/Christoph Schroeder);
- 2963199 : SWT print job (by Jonas Rüttimann);
- 2879650 : Path disclosure vulnerability in DisplayChart servlet;

Also fixed a rendering issue for polar charts using an inverted axis.

##### Version 1.0.14 (20-Nov-2011)
This release contains:

- support for multiple and logarithmic axes with PolarPlot;
- optional drop-shadows in plot rendering;
- fitting polynomial functions to a data series;
- some performance improvements in the TimeSeriesCollection class;
- mouse wheel rotation of pie charts;
- improved Maven support.

###### Patches

- 3435734 : Fix lines overlapping item labels (by Martin Hoeller);
- 3421088 : Bugfix for misalignment in BoxAndWhiskerRenderer;
- 2952086 : Enhancement for finding bounds in XYZDatasets;
- 2954302 : CategoryPointerAnnotation line calculation;
- 2902842 : HistogramDataset.addSeries() fires change change event (by Thomas A Caswell);
- 2868608 : Whisker width attribute for BoxAndWhiskerRenderer (by Peter Becker);
- 2868585 : Added useOutlinePaint flag for BoxAndWhiskerRenderer (by Peter Becker);
- 2850344 : PolarPlot enhancements (by Martin Hoeller);
- 2795746 : Support for polynomial regression;
- 2791407 : Fixes for findRangeBounds() in various renderers.

###### Bug Fixes

- 3440237 : Shadows always visible;
- 3432721 : PolarPlot doesn't work with logarithmic axis;
- 3433405 : LineChart3D - Problem with Item Labels;
- 3429707 : LogAxis endless loop;
- 3428870 : Missing argument check in TextAnnotation;
- 3418287 : RelativeDateFormatTest.java is locale dependent;
- 3353913 : Localisation fixes for ChartPanel, CompassPlot and PiePlot3D;
- 3237879 : RingPlot should respect getSectionOutlineVisible();
- 3190615 : Added missing clear() method in CategoryTableXYDataset;
- 3165708 : PolarChartPanel localisation fix;
- 3072674 : Bad handling of NaN in DefaultStatisticalCategoryDataset;
- 3035289 : StackedXYBarRenderer should respect series/item visible flags;
- 3026341 : Check for null in getDomain/RangeBounds() for XYShapeRenderer;
- 2947660 : AbstractCategoryRenderer fix null check in getLegendItems();
- 2946521 : StandardDialScale check majorTickIncrement argument;
- 2876406 : TimeTableXYDataset should support Comparable for series keys;
- 2868557 : BoxAndWhiskerRenderer should fire change event in setMedianVisible();
- 2849731 : For IntervalCategoryDataset and IntervalXYDataset, fix iterateRangeBounds() in DatasetUtilities;
- 2840132 : AbstractXYItemRenderer drawAnnotations doesn't set renderer index;
- 2810220 : Offset problem in StatisticalBarRenderer;
- 2802014 : Dial value border too small;
- 2781844 : XYPointerAnnotation arrow drawing;
- 1937486 : AreaRenderer doesn't respect AreaRendererEndType.LEVEL;

Also fixed:
- use of simple label offset in PiePlot;
- cached minY and maxY in TimeSeries.createCopy();
- scaling issues for charts copied to the clipboard;
- use of timezone in TimeTableXYDataset constructor;
- duplicate series names in XYSeriesCollection;
- HistogramDataset fires a change event in addSeries();
- check visibility of main chart title before drawing it;
- fixed serialization of PowerFunction2D, NormalDistributionFunction2D, and LineFunction2D;
- item label positioning for the AreaRenderer class when the plot has an horizontal orientation.

##### Version 1.0.13 (17-Apr-2009)

> SPECIAL NOTICE:  There will be a birds-of-a-feather session for JFreeChart at this year's JavaOne conference in San Francisco.  The session is scheduled for 6.45pm to 7.35pm on Wednesday 3 June.                        

This release contains:

- updates to the ChartPanel class to support copying charts to the clipboard, 
  panning and mouse-wheel zooming, and an overlay mechanism that supports
  crosshairs;
- enhancements to the auto-range calculation for axes, providing the ability
  to use subranges only and also to skip hidden series;
- updates for many of the CategoryItemRenderer implementations to ensure that
  they respect the 'seriesVisible' flags;
- an improvement to the TimeSeries class so that it is no longer necessary to
  specify the time period type in the constructor;
- a new SamplingXYLineRenderer for improving the performance of time series
  charts with large datasets;
- the XYSeries/XYSeriesCollection classes now cache the minimum and maximum
  data values to improve the performance of charts with large datasets;
- entities are now created for the chart, data area and axes, allowing mouse
  clicks to be detected for these regions;
- added a bar alignment factor to the XYBarRenderer class;
- a new 'errorIndicatorStroke' field for the StatisticalLineAndShapeRenderer 
  and XYErrorRenderer classes;
- added a new HeatMapDataset interface, DefaultHeatMapDataset implementation,
  and a HeatMapUtilities class to make it easier to create heat map charts;
- there is a new flag to allow an XYDataImageAnnotation to be included in the
  automatic range calculation for the axes;
- additional attributes in the XYTextAnnotation class;
- added a sampleFunction2DToSeries() method to the DatasetUtilities class;
- some changes to the ChartPanel class that help to work around a regression in
  JRE 1.6.0_10 relating to drawing in XOR mode.  Regarding this final point:
     * the default value for the useBuffer flag has changed to true, which means
       that all charts will, by default, be rendered into an off-screen image
       before being displayed in the ChartPanel;
     * the zoom rectangle is drawn using XOR mode *only* when the useBuffer
       flag has been set to false.
  For most usage, this should improve performance (but at the cost of using more
  memory for each ChartPanel used in your application);

###### Bug Fixes

- 2690293 : Problem with Javascript escape characters;
- 2617557 : StandardDialScale ignored tickLabelPaint;
- 2612649 : Stroke selection in plot property editor;
- 2583891 : SWTGraphics2D.fillPolygon() not implemented;
- 2564636 : Month constructor ignores Locale;
- 2502355 : ChartPanel sending multiple events;
- 2490803 : PeriodAxis.setRange() method doesn't take into account that the axis
          displays whole periods;

In addition, a bug in the SlidingCategoryDataset class has been fixed, the
correct outline paint is now used by GradientXYBarPainter, a new method
has been added to the ImageMapUtilities class to escape special characters
in Javascript strings to avoid problems with the OverLIB and DynamicDrive
tooltips, and there are some important fixes in the LogAxis class.

This release passes 2110 JUnit tests (0 failures) on JRE 1.6.0_12.


##### Version 1.0.12 (31-Dec-2008)

This release adds 
- support for minor tick marks;
- mapping datasets to more than one axis;
- an important fix for the XYSeries class (relating to the addOrUpdate() method);
- plus numerous other bug fixes.

This release passes 1996 JUnit test (0 failures) on JRE 1.6.0_10.

###### API Adjustments
- CategoryPlot : added mapDatasetToDomainAxes() and mapDatasetToRangeAxes() methods;
- Month : added a new constructor Month(Date, TimeZone, Locale) and deprecated Month(Date, TimeZone);
- Quarter : added a new constructor Quarter(Date, TimeZone, Locale) and deprecated Quarter(Date, TimeZone);
- XYPlot : added mapDatasetToDomainAxes() and mapDatasetToRangeAxes() methods;
- Year : added a new constructor Year(Date, TimeZone, Locale) and deprecated Year(Date, TimeZone);

###### Bug Fixes
- 2471906 : XYAreaRenderer with dashed outline - performance problem;
- 2452078 : StackedAreaChart has gaps;
- 2275695 : NullPointerException for SubCategoryAxis on plot with null dataset;
- 2221495 : XYLineAnnotation with dashed stroke;
- 2216511 : SWTBarChartDemo1 throws RuntimeException;
- 2201869 : DateAxis tick label position error;
- 2121818 : Label link lines for very thin RingPlot;
- 2113627 : XYStepRenderer item labels;
- 1955483 : XYSeries.addOrUpdate() problem.

Also fixed StackedXYBarRenderer which was ignoring the shadowsVisible attribute.


##### Version 1.0.11 (18-Sep-2008)

This release features:
- a new chart theming mechanism to allow charts to be restyled conveniently;
- a new BarPainter mechanism to enhance the appearance of bar charts;
- a new XYShapeRenderer class;
- a scaling facility for the XYDrawableAnnotation for drawing images within specific data coordinates;
- some new classes (XYTaskDataset, XYDataImageAnnotation and XYTitleAnnotation);
- a modification to the Year class to support an extended range; 
- various bug fixes and API improvements.  

There is an important bug fix for the StackedBarRenderer3D class (see bug 2031407).

This release passes 1,961 JUnit tests (0 failures) on JRE 1.6.0_07.

###### API Adjustments
- AbstractRenderer - added clearSeriesPaints() and clearSeriesStrokes() methods;
- BarRenderer - added shadowPaint attribute;
- CategoryAxis - added getCategoryMiddle() method;
- CategoryPlot - added getRendererCount() method;
- ChartFactory - added get/setChartTheme() methods;

NEWS
JFreeChart 1.0.19
-----------------
31 July 2014

This maintenance release corrects some issues in the new JavaFX support that
was added in the 1.0.18 release, as well as some general issues.

Bug Fixes:

- fixed clipping issues for combined plots in JavaFX;
- fix a memory leak in the new JavaFX ChartCanvas class;
- CombinedDomainXYPlot and CombinedRangeXYPlot now take into account the
  pannable flags in the subplots;
- FastScatterPlot panning direction is corrected;
- added rendering hints to sharpen gridlines and borders in most output formats;
- JFreeSVG updated to version 2.0.

In this release, we have also provided a preview of JSFreeChart, a free 2D
chart library written in JavaScript that is conceptually similar to JFreeChart
but runs directly in browsers.


JFreeChart 1.0.18
-----------------
3 July 2014

Added JavaFX support with a new ChartViewer control that supports tooltips,
mouse clicks, panning, zooming and a context menu.  Modified LogAxis to 
support super-scripted labels, improved the tick label generation for
NumberAxis, added a new ProbabilityAxis (patch from John St. Ledger), 
enhanced XYStepAreaRenderer to support a 'stepPoint' attribute (patch from 
Lukasz Rzeszotarski), provided configurable direction labels in 
CompassFormat (patch from Simon Legner), added methods to add data item 
objects to VectorSeries, XIntervalSeries, YIntervalSeries and 
XYIntervalSeries (for consistency with other data series objects), 
improved the TimeSeries range calculations, provided center text support 
for the RingPlot class, and updated JCommon to version 1.0.22 (this fixes 
a bug in pie label rendering).

Bug Fixes:

- BarRenderer3D now observes the 'visibleSeries' settings;
- TimeSeriesCollection now takes xRange into account when calculating the
range bounds;
- added missing localisation for 'Save As' option on ChartPanel;
- inverted SymbolAxis not displaying bands correctly;
- fixed fillDomainGridBand() method in AbstractXYItemRenderer;
- replaced ObjectList with java.util.Map to avoid bug in equals() methods.


JFreeChart 1.0.17
-----------------
24 November 2013

Enhanced XYSplineRenderer with new area fill (contributed by Klaus Rheinwald),
added a notify flag to all datasets that extend AbstractDataset, extended 
TimeSeriesCollection to validate TimeSeries keys for uniqueness, added a new
DirectionalGradientPaintTransformer (by Peter Kolb), updated OHLCSeries, added 
HMSNumberFormat, updated JCommon to version 1.0.21 (includes rotated text 
improvements) and fixed some minor bugs.

Bug Fixes:

 977 : Multithreading issue with DateAxis;
1084 : BorderArrangement.add() possible ClassCastException;
1099 : XYSeriesCollection.removeSeries(int) does not deregister listener;
1109 : WaterfallBarRenderer uses wrong color for diff 0.

And now a word from our sponsors, Object Refinery Limited...

> ORSON CHARTS 3D
> ---------------
> We are pleased to announce the release of Orson Charts, a 3D chart library
> for the Java(tm) platform.  Orson Charts has a similar architecture to 
> JFreeChart and provides high-quality 3D chart rendering using just the Java2D
> API.  The built-in Swing-based viewer provides zooming and full 360 degree
> rotation of charts in any direction.  Export options include PNG, PDF and SVG.
> You can download a free evaluation copy of Orson Charts from:
>
>    http://www.object-refinery.com/orsoncharts
>
> We've also added a tab to the JFreeChart demo application so you can see 
> Orson Charts in action.  Orson Charts is not free, but provides a superb
> complement to the extensive 2D charting capabilities of JFreeChart.


JFreeChart 1.0.16
-----------------
13 September 2013

*** THIS RELEASE REQUIRES JDK/JRE 1.6.0 OR LATER. ***

Provided subscript/superscript support for axis labels (via AttributedString),
new axis label positioning options, simplified ChartFactory methods,
added new methods to DatasetUtilities to interpolate y-values in XYDatasets,
added URLs to labels on CategoryAxis, seamless integration with 
JFreeSVG (http://www.jfree.org/jfreesvg/) and OrsonPDF 
(http://www.object-refinery.com/pdf/), improved the consistency of the 
SWTGraphics2D implementation, and fixed various bugs.  All the JUnit tests 
have been upgraded to JUnit 4.

Bug Fixes:
1107 : Fixed TimeZone issue in PeriodAxis.

Also fixed a line drawing issue with the StackedXYAreaRenderer, and a memory 
leak in the SWT ChartComposite class.

IMPORTANT: Funding is vital to the continued success of the JFreeChart project.
Please ask/persuade/beg your company to support us buy purchasing the
JFreeChart Developer Guide and demo source code (site license options are 
available for big companies)...and also please take a look at JFreeSVG and
OrsonPDF, useful products that we are charging money for!


JFreeChart 1.0.15
-----------------
4-July-2013

Added support for non-visible series in XYBarRenderer, minor gridlines in 
PolarPlot, legend item ordering, chart editor enhancements, updates to 
StandardDialScale, localisation files for Japanese and refactored parameter
checks.  

This release also fixes a minor security flaw in the DisplayChart class, 
detected and reported by OSI Security:

http://www.osisecurity.com.au/advisories/jfreechart-path-disclosure

Patches:

3500621 : LegendTitle order attribute (by Simon Kaczor);
3463807 : ChartComposite does not dispose popup (by Sebastiao Correia);
3204823 : PaintAlpha for 3D effects (by Dave Law);

Bug Fixes:

3561093 : Rendering anomaly for XYPlots;
3555275 : ValueAxis.reserveSpace() problem for axes with fixed dimension;
3521736 : DeviationRenderer optimisation (by Milan Ramaiya);
3514487 : SWTGraphics2D get/setStroke() problem;
3508799 : DefaultPolarItemRenderer does not populate seriesKey in LegendItem;
3482106 : Missing text in SWTGraphics2D (by Kevin Xu);
3484408 : Maven fixes (Martin Hoeller);
3484403 : DateAxis endless loop (by Martin Hoeller);
3446965 : TimeSeries calculates range incorrectly in addOrUpdate();

source/org/jfree/data/time/Day.java
/* ===========================================================
 * JFreeChart : a free chart library for the Java(tm) platform
 * ===========================================================
 *
 * (C) Copyright 2000-2016, by Object Refinery Limited and Contributors.
 *
 * Project Info:  http://www.jfree.org/jfreechart/index.html
 *
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation; either version 2.1 of the License, or
 * (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301,
 * USA.
 *
 * [Oracle and Java are registered trademarks of Oracle and/or its affiliates. 
 * Other names may be trademarks of their respective owners.]
 *
 * --------
 * Day.java
 * --------
 * (C) Copyright 2001-2016, by Object Refinery Limited.
 *
 * Original Author:  David Gilbert (for Object Refinery Limited);
 * Contributor(s):   -;
 *
 * Changes
 * -------
 * 11-Oct-2001 : Version 1 (DG);
 * 15-Nov-2001 : Updated Javadoc comments (DG);
 * 04-Dec-2001 : Added static method to parse a string into a Day object (DG);
 * 19-Dec-2001 : Added new constructor as suggested by Paul English (DG);
 * 29-Jan-2002 : Changed getDay() method to getSerialDate() (DG);
 * 26-Feb-2002 : Changed getStart(), getMiddle() and getEnd() methods to
 *               evaluate with reference to a particular time zone (DG);
 * 19-Mar-2002 : Changed the API for the TimePeriod classes (DG);
 * 29-May-2002 : Fixed bug in equals method (DG);
 * 24-Jun-2002 : Removed unnecessary imports (DG);
 * 10-Sep-2002 : Added getSerialIndex() method (DG);
 * 07-Oct-2002 : Fixed errors reported by Checkstyle (DG);
 * 10-Jan-2003 : Changed base class and method names (DG);
 * 13-Mar-2003 : Moved to com.jrefinery.data.time package, and implemented
 *               Serializable (DG);
 * 21-Oct-2003 : Added hashCode() method (DG);
 * 30-Sep-2004 : Replaced getTime().getTime() with getTimeInMillis() (DG);
 * 04-Nov-2004 : Reverted change of 30-Sep-2004, because it won't work for
 *               JDK 1.3 (DG);
 * ------------- JFREECHART 1.0.x ---------------------------------------------
 * 05-Oct-2006 : Updated API docs (DG);
 * 06-Oct-2006 : Refactored to cache first and last millisecond values (DG);
 * 16-Sep-2008 : Deprecated DEFAULT_TIME_ZONE (DG);
 * 02-Mar-2009 : Added new constructor with Locale (DG);
 * 05-Jul-2012 : Replaced getTime().getTime() with getTimeInMillis() (DG);
 * 03-Jul-2013 : Use ParamChecks (DG);
 *
 */

package org.jfree.data.time;

import java.io.Serializable;
import java.text.DateFormat;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.Locale;
import java.util.TimeZone;
import org.jfree.chart.util.ParamChecks;

import org.jfree.date.SerialDate;

/**
 * Represents a single day in the range 1-Jan-1900 to 31-Dec-9999.  This class
 * is immutable, which is a requirement for all {@link RegularTimePeriod}
 * subclasses.
 */
public class Day extends RegularTimePeriod implements Serializable {

    /** For serialization. */
    private static final long serialVersionUID = -7082667380758962755L;

    /** A standard date formatter. */
    protected static final DateFormat DATE_FORMAT
            = new SimpleDateFormat("yyyy-MM-dd");

    /** A date formatter for the default locale. */
    protected static final DateFormat DATE_FORMAT_SHORT 
            = DateFormat.getDateInstance(DateFormat.SHORT);

    /** A date formatter for the default locale. */
    protected static final DateFormat DATE_FORMAT_MEDIUM 
            = DateFormat.getDateInstance(DateFormat.MEDIUM);

    /** A date formatter for the default locale. */
    protected static final DateFormat DATE_FORMAT_LONG 
            = DateFormat.getDateInstance(DateFormat.LONG);

    /** The day (uses SerialDate for convenience). */
    private SerialDate serialDate;

    /** The first millisecond. */
    private long firstMillisecond;

    /** The last millisecond. */
    private long lastMillisecond;

    /**
     * Creates a new instance, derived from the system date/time (and assuming
     * the default timezone).
     */
    public Day() {
        this(new Date());
    }

    /**
     * Constructs a new one day time period.
     *
     * @param day  the day-of-the-month.
     * @param month  the month (1 to 12).
     * @param year  the year (1900 &lt;= year &lt;= 9999).
     */
    public Day(int day, int month, int year) {
        this.serialDate = SerialDate.createInstance(day, month, year);
        peg(Calendar.getInstance());
    }

    /**
     * Constructs a new one day time period.
     *
     * @param serialDate  the day ({@code null} not permitted).
     */
    public Day(SerialDate serialDate) {
        ParamChecks.nullNotPermitted(serialDate, "serialDate");
        this.serialDate = serialDate;
        peg(Calendar.getInstance());
    }

    /**
     * Constructs a new instance, based on a particular date/time and the
     * default time zone.
     *
     * @param time  the time ({@code null} not permitted).
     *
     * @see #Day(Date, TimeZone)
     */
    public Day(Date time) {
        // defer argument checking...
        this(time, TimeZone.getDefault(), Locale.getDefault());
    }

    /**
     * Constructs a new instance, based on a particular date/time and time zone.
     *
     * @param time  the date/time.
     * @param zone  the time zone.
     *
     * @deprecated As of 1.0.13, use the constructor that specifies the locale
     *     also.
     */
    public Day(Date time, TimeZone zone) {
        this(time, zone, Locale.getDefault());
    }

    /**
     * Constructs a new instance, based on a particular date/time and time zone.
     *
     * @param time  the date/time ({@code null} not permitted).
     * @param zone  the time zone ({@code null} not permitted).
     * @param locale  the locale ({@code null} not permitted).
     */
    public Day(Date time, TimeZone zone, Locale locale) {
        ParamChecks.nullNotPermitted(time, "time");
        ParamChecks.nullNotPermitted(zone, "zone");
        ParamChecks.nullNotPermitted(locale, "locale");
        Calendar calendar = Calendar.getInstance(zone, locale);
        calendar.setTime(time);
        int d = calendar.get(Calendar.DAY_OF_MONTH);
        int m = calendar.get(Calendar.MONTH) + 1;
        int y = calendar.get(Calendar.YEAR);
        this.serialDate = SerialDate.createInstance(d, m, y);
        peg(calendar);
    }

    /**
     * Returns the day as a {@link SerialDate}.  Note: the reference that is
     * returned should be an instance of an immutable {@link SerialDate}
     * (otherwise the caller could use the reference to alter the state of
     * this <code>Day</code> instance, and <code>Day</code> is supposed
     * to be immutable).
     *
     * @return The day as a {@link SerialDate}.
     */
    public SerialDate getSerialDate() {
        return this.serialDate;
    }

    /**
     * Returns the year.
     *
     * @return The year.
     */
    public int getYear() {
        return this.serialDate.getYYYY();
    }

    /**
     * Returns the month.
     *
     * @return The month.
     */
    public int getMonth() {
        return this.serialDate.getMonth();
    }

    /**
     * Returns the day of the month.
     *
     * @return The day of the month.
     */
    public int getDayOfMonth() {
        return this.serialDate.getDayOfMonth();
    }

    /**
     * Returns the first millisecond of the day.  This will be determined
     * relative to the time zone specified in the constructor, or in the
     * calendar instance passed in the most recent call to the
     * {@link #peg(Calendar)} method.
     *
     * @return The first millisecond of the day.
     *
     * @see #getLastMillisecond()
     */
    @Override
    public long getFirstMillisecond() {
        return this.firstMillisecond;
    }

    /**
     * Returns the last millisecond of the day.  This will be
     * determined relative to the time zone specified in the constructor, or
     * in the calendar instance passed in the most recent call to the
     * {@link #peg(Calendar)} method.
     *
     * @return The last millisecond of the day.
     *
     * @see #getFirstMillisecond()
     */
    @Override
    public long getLastMillisecond() {
        return this.lastMillisecond;
    }

    /**
     * Recalculates the start date/time and end date/time for this time period
     * relative to the supplied calendar (which incorporates a time zone).
     *
     * @param calendar  the calendar ({@code null} not permitted).
     *
     * @since 1.0.3
     */
    @Override
    public void peg(Calendar calendar) {
        this.firstMillisecond = getFirstMillisecond(calendar);
        this.lastMillisecond = getLastMillisecond(calendar);
    }

    /**
     * Returns the day preceding this one.
     *
     * @return The day preceding this one.
     */
    @Override
    public RegularTimePeriod previous() {
        Day result;
        int serial = this.serialDate.toSerial();
        if (serial > SerialDate.SERIAL_LOWER_BOUND) {
            SerialDate yesterday = SerialDate.createInstance(serial - 1);
            return new Day(yesterday);
        }
        else {
            result = null;
        }
        return result;
    }

    /**
     * Returns the day following this one, or {@code null} if some limit
     * has been reached.
     *
     * @return The day following this one, or {@code null} if some limit
     *         has been reached.
     */
    @Override
    public RegularTimePeriod next() {
        Day result;
        int serial = this.serialDate.toSerial();
        if (serial < SerialDate.SERIAL_UPPER_BOUND) {
            SerialDate tomorrow = SerialDate.createInstance(serial + 1);
            return new Day(tomorrow);
        }
        else {
            result = null;
        }
        return result;
    }

    /**
     * Returns a serial index number for the day.
     *
     * @return The serial index number.
     */
    @Override
    public long getSerialIndex() {
        return this.serialDate.toSerial();
    }

    /**
     * Returns the first millisecond of the day, evaluated using the supplied
     * calendar (which determines the time zone).
     *
     * @param calendar  calendar to use ({@code null} not permitted).
     *
     * @return The start of the day as milliseconds since 01-01-1970.
     *
     * @throws NullPointerException if <code>calendar</code> is
     *     {@code null}.
     */
    @Override
    public long getFirstMillisecond(Calendar calendar) {
        int year = this.serialDate.getYYYY();
        int month = this.serialDate.getMonth();
        int day = this.serialDate.getDayOfMonth();
        calendar.clear();
        calendar.set(year, month - 1, day, 0, 0, 0);
        calendar.set(Calendar.MILLISECOND, 0);
        return calendar.getTimeInMillis();
    }

    /**
     * Returns the last millisecond of the day, evaluated using the supplied
     * calendar (which determines the time zone).
     *
     * @param calendar  calendar to use ({@code null} not permitted).
     *
     * @return The end of the day as milliseconds since 01-01-1970.
     *
     * @throws NullPointerException if <code>calendar</code> is
     *     {@code null}.
     */
    @Override
    public long getLastMillisecond(Calendar calendar) {
        int year = this.serialDate.getYYYY();
        int month = this.serialDate.getMonth();
        int day = this.serialDate.getDayOfMonth();
        calendar.clear();
        calendar.set(year, month - 1, day, 23, 59, 59);
        calendar.set(Calendar.MILLISECOND, 999);
        return calendar.getTimeInMillis();
    }

    /**
     * Tests the equality of this Day object to an arbitrary object.  Returns
     * true if the target is a Day instance or a SerialDate instance
     * representing the same day as this object. In all other cases,
     * returns false.
     *
     * @param obj  the object ({@code null} permitted).
     *
     * @return A flag indicating whether or not an object is equal to this day.
     */
    @Override
    public boolean equals(Object obj) {
        if (obj == this) {
            return true;
        }
        if (!(obj instanceof Day)) {
            return false;
        }
        Day that = (Day) obj;
        if (!this.serialDate.equals(that.getSerialDate())) {
            return false;
        }
        return true;
    }

    /**
     * Returns a hash code for this object instance.  The approach described by
     * Joshua Bloch in "Effective Java" has been used here:
     * <p>
     * <code>http://developer.java.sun.com/developer/Books/effectivejava
     * /Chapter3.pdf</code>
     *
     * @return A hash code.
     */
    @Override
    public int hashCode() {
        return this.serialDate.hashCode();
    }

    /**
     * Returns an integer indicating the order of this Day object relative to
     * the specified object:
     *
     * negative == before, zero == same, positive == after.
     *
     * @param o1  the object to compare.
     *
     * @return negative == before, zero == same, positive == after.
     */
    @Override
    public int compareTo(Object o1) {
        int result;

        // CASE 1 : Comparing to another Day object
        // ----------------------------------------
        if (o1 instanceof Day) {
            Day d = (Day) o1;
            result = -d.getSerialDate().compare(this.serialDate);
        }

        // CASE 2 : Comparing to another TimePeriod object
        // -----------------------------------------------
        else if (o1 instanceof RegularTimePeriod) {
            // more difficult case - evaluate later...
            result = 0;
        }

        // CASE 3 : Comparing to a non-TimePeriod object
        // ---------------------------------------------
        else {
            // consider time periods to be ordered after general objects
            result = 1;
        }

        return result;
    }

    /**
     * Returns a string representing the day.
     *
     * @return A string representing the day.
     */
    @Override
    public String toString() {
        return this.serialDate.toString();
    }

    /**
     * Parses the string argument as a day.
     * <P>
     * This method is required to recognise YYYY-MM-DD as a valid format.
     * Anything else, for now, is a bonus.
     *
     * @param s  the date string to parse.
     *
     * @return {@code null} if the string does not contain any parseable
     *      string, the day otherwise.
     */
    public static Day parseDay(String s) {
        try {
            return new Day (Day.DATE_FORMAT.parse(s));
        }
        catch (ParseException e1) {
            try {
                return new Day (Day.DATE_FORMAT_SHORT.parse(s));
            }
            catch (ParseException e2) {
              // ignore
            }
        }
        return null;
    }

}
